### PR TITLE
HPCC-25722 Allow annotations on Roxie services

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [x] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: helm/hpcc/templates/_helpers.tpl [hunk 1]
```diff
@@ -763,39 +763,44 @@ Generate list of available services
...
+  port: {{ $service.servicePort }}
   target: {{ $roxie.name }}
-  public: {{ $service.external }}
+  public: {{ $service.public }}
    {{- end -}}
   {{- end }}
 {{ end -}}
 {{- end -}}
 {{- range $esp := $.Values.esp -}}
+{{- if not (hasKey $esp "service") -}}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:778]: out of curiosity, are there esp services without a public/internal service?

File: helm/hpcc/templates/esp.yaml [hunk 2]
```diff
@@ -73,6 +73,12 @@ data:
 {{- if not .disabled -}}
 {{- $env := concat ($.Values.global.env | default list) (.env | default list) -}}
 {{- $application := .application | default "eclwatch" -}}
+{{- if not (hasKey . "service") -}}
+{{- $_ := set . "service" (dict "port" (.port | default 0) "servicePort" (.servicePort | default 0) "public" (.public | default false) ) -}}
+{{- $_ := unset . "port" -}}
+{{- $_ := unset . "servicePort" -}}
+{{- $_ := unset . "public" -}}
```
Comment (jakesmith) [inline helm/hpcc/templates/esp.yaml:80]: is above for backward compatibility, for old value yamls that don't define the new .service.servicePort etc. spec?
If so, need similar for Sasha and Roxie to support their old format?

File: helm/hpcc/templates/esp.yaml [hunk 3]
```diff
@@ -135,15 +141,23 @@ apiVersion: v1
...
   ports:
-  - port: {{ required "servicePort must be specified" .servicePort }}
+  - port: {{ required "servicePort must be specified" .service.servicePort }}
     protocol: TCP
-    targetPort: {{ .port | default 8880 }}
+    targetPort: {{ .service.port | default 8880 }}
   selector:
     run: {{ .name | quote }}
-  type: {{ .public | ternary "LoadBalancer" "ClusterIP" }}
+  type: {{ .service.public | ternary "LoadBalancer" "ClusterIP" }}
```
Comment (jakesmith) [inline helm/hpcc/templates/esp.yaml:159]: The whole of this Service generation before/now looks very generic, only thing that's specific to esp perhaps is the default ports.
It would be good to move it into _helpers.tpl, as a generateService call, and use in other places too.
(think it's same for example as on in Roxie apart from the default port aspect)

File: helm/hpcc/templates/localroxie.yaml [hunk 4]
```diff
@@ -121,21 +121,28 @@ spec:
...
   ports:
-  - port: {{ $service.port }}
+  - port: {{ $service.servicePort }}
     protocol: TCP
-    targetPort: {{ $service.port }}
+    targetPort: {{ $service.servicePort }}
   selector:
     roxie-server: {{ $servername | quote }}
-  type: {{ if $service.external -}} LoadBalancer {{- else -}} ClusterIP {{- end }}
+  type: {{ if $service.public -}} LoadBalancer {{- else -}} ClusterIP {{- end }}
```
Comment (jakesmith) [inline helm/hpcc/templates/localroxie.yaml:145]: see comment in esp.yaml - looks like generate of a Service is very similar, could be commoned up.

File: helm/hpcc/values.yaml [hunk 5]
```diff
@@ -250,8 +250,7 @@ dali:
 - name: mydali
   services: # internal house keeping services
     coalescer:
-      #servicePort: 8877
-      #interval: 2 # (hours)
+      interval: 2 # (hours)
```
Comment (jakesmith) [inline helm/hpcc/values.yaml:253]: intentionally commented in?

File: helm/hpcc/values.yaml [hunk 6]
```diff
@@ -300,14 +301,18 @@ sasha:
     #throttle: 0 # throttle ratio (0-99, 0 no throttling, 50 is half speed)
 
   dfurecovery-archiver:
-    #disabled: true
+    disabled: true
+    service:
```
Comment (jakesmith) [inline helm/hpcc/values.yaml:305]: I assume not intended, just disabled while testing?

File: helm/hpcc/values.yaml [hunk 7]
```diff
@@ -300,14 +301,18 @@ sasha:
...
+      servicePort: 8877
     #limit: 20 # threshold number of DFU recovery items before archiving starts (0 disables)
     #cutoff: 4 # minimum DFU recovery item age to archive (days)
     #interval: 12 # minimum interval between running DFU recovery archiver(in hours, 0 disables)
     #at: "* * * * *" # schedule to run DFU recovery archiver (cron format)
 
   file-expiry:
-    #disabled: true
+    disabled: true
```
Comment (jakesmith) [inline helm/hpcc/values.yaml:313]: I assume not intended, just disabled while testing?

File: helm/hpcc/templates/esp.yaml (hunk 2 above)
Comment (richardkchapman) [reply helm/hpcc/templates/esp.yaml:80]: Yes. Not sure whether it's a good idea or not, or whether should just require people to change. Given the schema changes I suspect this code may be moot anyway.

File: roxie/ccd/ccdmain.cpp [hunk 8]
```diff
@@ -739,7 +739,7 @@ int CCD_API roxie_main(int argc, const char *argv[], const char * defaultYaml)
                     // Makes debugging easier...
                     IPropertyTree *service = topology->addPropTree("services");
                     service->setProp("@name", "query");
-                    service->setPropInt("@port", 9876);
+                    service->setPropInt("@servicePort", 9876);
```
Comment (jakesmith) [inline roxie/ccd/ccdmain.cpp:742]: is this correct? Shouldn't it be listening to @port (the internally defined port to use), which should default to @servicePort, if @port was not defined in values yaml

Comment (jakesmith) [reply roxie/ccd/ccdmain.cpp:742]: or maybe this isn't being used to listen to here,
but is being used contact the exposed service ?

File: helm/hpcc/templates/esp.yaml (hunk 2 above)
Comment (jakesmith) [reply helm/hpcc/templates/esp.yaml:80]: yeah, given the schema, I think no point, though disruptive to users potentially.

If this code is removed though, I think there is other code that node checks .service.X - that will need to be careful to check if key "service" exists first.

File: helm/hpcc/templates/localroxie.yaml [hunk 9]
```diff
@@ -121,21 +121,28 @@ spec:
...
+  annotations:
+{{ toYaml $service.annotations | indent 4 }}
+{{ end -}}
 spec:
   ports:
-  - port: {{ $service.port }}
+  - port: {{ $service.servicePort }}
     protocol: TCP
-    targetPort: {{ $service.port }}
+    targetPort: {{ $service.servicePort }}
```
Comment (jakesmith) [inline helm/hpcc/templates/localroxie.yaml:142]: shouldn't this remain as port?

Comment (richardkchapman) [reply helm/hpcc/templates/localroxie.yaml]: Roxie does not allow them to be different. I'm not sure why other components do, TBH.

File: roxie/ccd/ccdmain.cpp (hunk 8 above)
Comment (richardkchapman) [reply roxie/ccd/ccdmain.cpp:742]: Roxie uses a single port to specify both internal and external - they are not allowed to differ.

File: helm/hpcc/values.yaml (hunk 7 above)
Comment (richardkchapman) [reply helm/hpcc/values.yaml]: reverted

File: helm/hpcc/values.yaml (hunk 6 above)
Comment (richardkchapman) [reply helm/hpcc/values.yaml]: reverted

File: helm/hpcc/values.yaml (hunk 5 above)
Comment (richardkchapman) [reply helm/hpcc/values.yaml]: reverted

File: helm/hpcc/templates/localroxie.yaml (hunk 4 above)
Comment (richardkchapman) [reply helm/hpcc/templates/localroxie.yaml]: Done

File: helm/hpcc/templates/esp.yaml (hunk 3 above)
Comment (richardkchapman) [reply helm/hpcc/templates/esp.yaml]: Done

File: roxie/ccd/ccdmain.cpp [hunk 10]
```diff
@@ -1326,7 +1326,7 @@ int CCD_API roxie_main(int argc, const char *argv[], const char * defaultYaml)
                     unsigned numThreads = roxieFarm.getPropInt("@numThreads", 0);
                     if (!numThreads)
                         numThreads = numServerThreads;
-                    unsigned port = roxieFarm.getPropInt("@port", ROXIE_SERVER_PORT);
```
Comment (richardkchapman) [inline roxie/ccd/ccdmain.cpp:1329]: This breaks bare-metal...

Comment (jakesmith) [reply roxie/ccd/ccdmain.cpp:1329]: are other places ok to default in bare metal (e.g. line 1205) ?

Perhaps instead of conditional code here, check @port in bare-metal of farmers upfront, e.g. in block around line 656

File: helm/hpcc/values.schema.json [hunk 11]
```diff
@@ -953,11 +1005,29 @@
           "type": "integer",
           "default": 1
         },
-        "servicePort": {
-          "type": "integer"
-        },
-        "port": {
-          "type": "integer"
+        "service": {
```
Comment (jakesmith) [inline helm/hpcc/values.schema.json:1008]: I think this service section is identical to esp's apart from no "public" field - can it be commoned up?

File: roxie/ccd/ccdmain.cpp (hunk 10 above)
Comment (richardkchapman) [reply roxie/ccd/ccdmain.cpp:1329]: Others are inside #ifdef's that make them containerized only.

File: helm/hpcc/templates/_helpers.tpl [hunk 12]
```diff
@@ -945,23 +952,31 @@ Pass in dict with root and me
...
-  - port: {{ .me.servicePort }}
+  - port: {{ required "servicePort must be specified" .service.servicePort }}
     protocol: TCP
-    targetPort: {{ .port | default 8877 }}
+    targetPort: {{ .service.port | default .defaultPort }}
   selector:
-    run: {{ .selectorName | default $serviceName | quote }}
-  type: ClusterIP
+    server: {{ .selector | quote }}
+  type: {{ and ( hasKey .service "public") .service.public | ternary "LoadBalancer" "ClusterIP" }}
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:979]: Not for sasha...

File: helm/hpcc/values.schema.json (hunk 11 above)
Comment (richardkchapman) [reply helm/hpcc/values.schema.json:1008]: The lack of public is enough reason to not want to

**Review by ghalliday — COMMENTED** [review]
@richardkchapman looks good.  a few minor comments.

**Comment by richardkchapman** [comment]
"even if public was missing the hasKey test would not be needed since it would evaluate to nil/false."

You would think that, but ternary seems to object to the first arg not being boolean, rather than treating nil as false.

File: helm/hpcc/templates/_helpers.tpl [hunk 13]
```diff
@@ -763,39 +763,44 @@ Generate list of available services
...
   target: {{ $roxie.name }}
-  public: {{ $service.external }}
+  public: {{ $service.public }}
    {{- end -}}
   {{- end }}
 {{ end -}}
 {{- end -}}
 {{- range $esp := $.Values.esp -}}
+{{- if not (hasKey $esp "service") -}}
+{{- $_ := set $esp "service" (dict "port" (.port | default 0) "servicePort" (.servicePort | default 0) "public" (.public | default false) ) -}}
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:779]: Yes, though I think this change is now redundant since service is compulsory in the schema. I will remove it.

File: helm/hpcc/templates/_helpers.tpl [hunk 14]
```diff
@@ -804,11 +809,13 @@ Generate list of available services
 {{- if not $sashaServices.disabled -}}
 {{- range $sashaName, $_sasha := $sashaServices -}}
 {{- $sasha := ($_sasha | default dict) -}}
-{{- if and (not $sasha.disabled) ($sasha.servicePort) -}}
+{{- if and (not $sasha.disabled) (hasKey $sasha "service") -}}
+{{- if $sasha.service.servicePort -}}
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:813]: Sasha is a bit different as service is optional

File: helm/hpcc/templates/dali.yaml [hunk 15]
```diff
@@ -37,11 +37,13 @@ spec:
   selector:
     matchLabels:
       run: {{ $dali.name | quote }}
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply helm/hpcc/templates/dali.yaml:39]: I used a server: label for associating services with pods, rather than a run: label - they don't always match (e.g. sasha services that are done inside the dali, and roxie clusters where not every node is a server).

I left the run: label in place as that seems to be something of a helm convention, and has its uses for interrogating the state of the system via kubectl etc. I don't think we actually use it ourselves now though.

File: helm/hpcc/values.schema.json [hunk 16]
```diff
@@ -709,6 +734,42 @@
...
+              "description": "The local port used by the pod",
+              "default": 8880
+            },
+            "servicePort": {
+              "type": "integer",
+              "description": "The port that this service will be exposed on"
+            },
+            "visibility": {
+              "type": "string",
+              "description": "Should this service be exposed outside the cluster, locally or to the internet"
```
Comment (ghalliday) [inline helm/hpcc/values.schema.json:762]: "enum": [ "cluster", "local", "global" ],?

File: helm/hpcc/values.schema.json (hunk 11 above)
Comment (ghalliday) [reply helm/hpcc/values.schema.json:1041]: Does this change now you have local/global?

File: helm/hpcc/values.yaml [hunk 17]
```diff
@@ -63,6 +63,21 @@ global:
   #   postJobCommand: "kill $(pgrep linkerd2-proxy)"
   #   postJobCommandViaSidecar: true
 
+  ## visibilities section can be used to set labels, annotations and service type for any service with the specified visibility
+  visibilities:
+    cluster:
+      type: ClusterIP
+    local:
+      annotations:
+        service.beta.kubernetes.io/azure-load-balancer-internal: "true"
```
Comment (ghalliday) [inline helm/hpcc/values.yaml:72]: Should this be commented out by default?

File: helm/hpcc/values.yaml [hunk 18]
```diff
@@ -365,48 +383,59 @@ esp:
...
+  service:
+    ## port can be used to change the local port used by the pod. If omitted, the default port (8880) is used
+    port: 8888
+    ## servicePort controls the port that this service will be exposed on, either internally to the cluster, or externally
+    servicePort: 8010
+    ## Specify visibility: local (or global) if you want the service available from outside the cluster. Typically, eclwatch and wsecl are published externally, while eclservices is designed for internal use.
+    visibility: local
+    ## Annotations can be specified on a service - for example to specify provider-specific information such as service.beta.kubernetes.io/azure-load-balancer-internal-subnet
+    annotations:
+      service.beta.kubernetes.io/azure-load-balancer-internal-subnet: "mysubnet"
```
Comment (ghalliday) [inline helm/hpcc/values.yaml:395]: comment out? or remove entirely now you have your visibility section?  Same for the label below.

File: helm/hpcc/templates/_helpers.tpl [hunk 19]
```diff
@@ -945,23 +949,71 @@ Pass in dict with root and me
...
 
 {{/*
-A template to generate Sasha service
-Pass in dict me
+A template to generate the type of a service based on the visibility setting
+Pass in dict with .root, .visibility defined
 */}}
-{{- define "hpcc.addSashaService" }}
-{{- $serviceName := printf "sasha-%s" .me.name }}
+{{- define "hpcc.getVisibilityType" }}
```
Comment (ghalliday) [inline helm/hpcc/templates/_helpers.tpl:955]: Is this the correct name for the function?  getServiceType()?

**Review by ghalliday — COMMENTED** [review]
@richardkchapman I like the visibility section....  A few comments

File: helm/hpcc/values.schema.json (hunk 16 above)
Comment (richardkchapman) [reply helm/hpcc/values.schema.json:762]: It was originally, but it's now possible for people to add their own visibility classes if they want

File: helm/hpcc/templates/_helpers.tpl (hunk 19 above)
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:955]: It is mapping a visibility value to a service type value. I could rename it if you think a different name is less confusing. Though I suspect if I can get boolean returns working I will probably replace it with something like "isVisibilityExternal"

File: helm/hpcc/values.schema.json (hunk 11 above)
Comment (richardkchapman) [reply helm/hpcc/values.schema.json:1041]: I don't think so - we don't want to allow sasha services to be made visible.

File: helm/hpcc/values.yaml (hunk 17 above)
Comment (richardkchapman) [reply helm/hpcc/values.yaml:72]: No - because we want by default the system to work on azure without exposing things to the internet (thus keeping Flavio happy). It will be harmless on non-azure systems.

File: helm/hpcc/values.yaml (hunk 18 above)
Comment (richardkchapman) [reply helm/hpcc/values.yaml:395]: I should comment it out. The idea is that you can override on individual services as well as setting up visibility "classes" of service.

**Review by ghalliday — APPROVED** [review]
Looks ok.  Jake should also check.

File: helm/hpcc/values.schema.json (hunk 16 above)
Comment (ghalliday) [reply helm/hpcc/values.schema.json:762]: Should the restriction be removed for the other place where visibility is defined?

Comment (richardkchapman) [reply helm/hpcc/values.schema.json:762]: Yes - well spotted. I will push another fix

**Review by ghalliday — APPROVED** [review]
Looks good to me.

File: esp/smc/SMCLib/TpContainer.cpp [hunk 20]
```diff
@@ -500,7 +500,7 @@ void CTpWrapper::getServices(double version, const char* serviceType, const char
     {
         IPropertyTree& service = itr->query();
         //Only show the public services for now
-        if (!service.getPropBool("@public"))
+        if (!service.getPropBool("@public", false))
```
Comment (jakesmith) [inline esp/smc/SMCLib/TpContainer.cpp:503]: super trivial: we don't normally add the default param in cases like this.

File: helm/hpcc/templates/_helpers.tpl [hunk 21]
```diff
@@ -840,13 +840,13 @@ Generate list of available services
...
-   {{- if ne (int $service.port) 0 -}}
+   {{- if ne (int $service.servicePort) 0 -}}
 - name: {{ $service.name }}
   class: roxie
   type: roxie
-  port: {{ $service.port }}
+  port: {{ $service.servicePort }}
   target: {{ $roxie.name }}
-  public: {{ $service.external }}
+  public: {{ (ne ( include "hpcc.isVisibilityPublic" (dict "root" $ "visibility" $service.visibility))  "") | ternary "true" "false" }}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:849]: super trivial: extra space before ""

File: helm/hpcc/templates/_helpers.tpl [hunk 22]
```diff
@@ -855,24 +855,26 @@ Generate list of available services
...
   type: {{ $esp.application }}
-  port: {{ $esp.servicePort }}
+  port: {{ $esp.service.servicePort }}
   {{- if hasKey $esp "tls" }}
   tls: {{ $esp.tls }}
   {{- else }}
   tls: {{ ($.Values.certificates | default dict).enabled }}
   {{- end }}
-  public: {{ $esp.public }}
+  public: {{ (ne ( include "hpcc.isVisibilityPublic" (dict "root" $ "visibility" $esp.service.visibility))  "") | ternary "true" "false" }}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:864]: trivial: if isVisibilityPublic returned only "true" or "false" (never ""), could you simplify?

File: helm/hpcc/templates/_helpers.tpl [hunk 23]
```diff
@@ -1023,23 +1027,71 @@ Pass in dict with root and me
...
 {{/*
-A template to generate Sasha service
-Pass in dict me
+A template to generate the type of a service based on the visibility setting
+Pass in dict with .root, .visibility defined
 */}}
-{{- define "hpcc.addSashaService" }}
-{{- $serviceName := printf "sasha-%s" .me.name }}
+{{- define "hpcc.isVisibilityPublic" }}
+{{- if and (hasKey . "visibility") .visibility -}}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:1034]: isn't :
```
{{- if .visibility -}}
```

enough here?

File: helm/hpcc/templates/_helpers.tpl [hunk 24]
```diff
@@ -1023,23 +1027,71 @@ Pass in dict with root and me
...
+ {{- if hasKey .root.Values.global "visibilities" -}}
+  {{- if hasKey .root.Values.global.visibilities .visibility -}}
+   {{- $globalServiceInfo := get .root.Values.global.visibilities .visibility -}}
+   {{- if (not (eq $globalServiceInfo.type "ClusterIP")) -}}
+   true
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:1039]: I think really, should not indent here, i.e. not return "    true"

File: helm/hpcc/templates/_helpers.tpl [hunk 25]
```diff
@@ -1023,23 +1027,71 @@ Pass in dict with root and me
...
+   {{- end -}}
+  {{- else -}}
+   {{- required (printf "Specified service visibility %s not found in global visibilities section" .service.visibility) nil -}}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:1042]: trivia: is required ... nil , preferable to fatal .. ?

**Review by jakesmith — APPROVED** [review]
@richardkchapman - trivial comments only

File: helm/hpcc/templates/_helpers.tpl (hunk 22 above)
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:864]: No - because it would be returning them as strings, not as a boolean

File: helm/hpcc/templates/_helpers.tpl (hunk 24 above)
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:1039]: Doesn't matter given the -}} above. I decided it was more legible this way

File: helm/hpcc/templates/_helpers.tpl (hunk 25 above)
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:1042]: It seems to give a fairly nice error message. I can try fatal and see if it seems better

File: esp/smc/SMCLib/TpContainer.cpp (hunk 20 above)
Comment (richardkchapman) [reply esp/smc/SMCLib/TpContainer.cpp:503]: I was trying to revert the code to the way it was (and failed). I'll take it out

File: helm/hpcc/templates/_helpers.tpl (hunk 21 above)
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:849]: Will remove

File: helm/hpcc/templates/_helpers.tpl (hunk 23 above)
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:1034]: Maybe. But I tested it like this...

File: helm/hpcc/templates/_helpers.tpl (hunk 25 above)
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:1042]: I can't find any documentation on fatal

Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:1042]: (though I did find a typo in the error reporting line when I tried to test it so that was handy!)
